### DISCUSSION TITLE
Convenience Refinements for enums

### DISCUSSION THREAD

**Discussion by jamesarosen** [colinhacks/zod#2130] [Q&A] [answered]
I have a Zod native enum:
```ts
enum HttpMethod {
  DELETE = 'DELETE'
  GET = 'GET'
  HEAD = 'HEAD'
  OPTIONS = 'OPTIONS'
  POST = 'POST'
  PUT = 'PUT'
}

const httpMethod = z.nativeEnum(HttpMethod)
```

And then I use that in another zod parser:
```ts
const getRequest = z.object({
  method: z.literal(HttpMethod.GET)
})
```

This works fine, but it would be nice to be able to get a version of the `httpMethod` parser that only accepts `'GET'`. I can do it with a refinement:
```ts
const getRequest = z.object({
  method: httpMethod.refine(m => m === HttpMethod.GET)
})
```

I was just hoping for something shorter and more expressive, like
```ts
const getRequest = z.object({
  method: httpMethod.GET
})
```

**Comment by jamesarosen** [comment]
Similarly, it would be nice to be able to get the enum types out of a zod-defined enum more easily.

This compiles:
```ts
const httpMethod = z.enum(['DELETE', 'GET', 'HEAD', 'OPTIONS', 'POST', 'PUT'])
type HttpMethod = z.infer<typeof httpMethod>
type GetRequest = {
  method: HttpMethod[0]
}
const x: GetRequest = { method: 'POST' }
```
even though `{ method: 'POST' }` is not assignable to the type `{ method: 'GET' }`

This doesn't work:
```ts
const httpMethod = z.enum(['DELETE', 'GET', 'HEAD', 'OPTIONS', 'POST', 'PUT'])

type GetRequest = {
  method: httpMethod.enum.GET
}
```
TypeScript complains, `Cannot find namespace 'httpMethod'.`

Nor does this:
```ts
const httpMethod = z.enum(['DELETE', 'GET', 'HEAD', 'OPTIONS', 'POST', 'PUT'])
type HttpMethod = z.infer<typeof httpMethod>
type GetRequest = {
  method: HttpMethod['GET']
}
```
where TypeScript complains, `Property 'GET' does not exist on type '"DELETE" | "GET" | "HEAD" | "OPTIONS" | "POST" | "PUT"'.`

**Comment by JacobWeisenburger** [comment]
This works for me. What is not working about this for you?
```ts
enum HttpMethod {
    DELETE = 'DELETE',
    GET = 'GET',
    HEAD = 'HEAD',
    OPTIONS = 'OPTIONS',
    POST = 'POST',
    PUT = 'PUT',
}

const getRequest = z.object( {
    method: z.literal( HttpMethod.GET )
} )

console.log( getRequest.parse( { method: 'GET' } ) ) // { method: 'GET' }
```

---

You forgot to use `typeof` on `httpMethod.enum.GET`
```ts
const httpMethod = z.enum( [ 'DELETE', 'GET', 'HEAD', 'OPTIONS', 'POST', 'PUT' ] )
type HttpMethod = z.infer<typeof httpMethod>
type GetRequest = {
    method: typeof httpMethod.enum.GET
}
const x: GetRequest = { method: 'POST' }
//                      ^^^^^^
// Type '"POST"' is not assignable to type '"GET"'
```

Other Resources:
https://gist.github.com/JacobWeisenburger/1c72eda9e823cd6caed327db470f0613

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by jamesarosen** [reply]
Thank you for chiming in, Jacob!

Everything you've said is correct. I think a longer example may better describe what I want.

We'll start with a version based on your code. This works:
```ts
enum HttpMethod {
  DELETE = 'DELETE'
  GET = 'GET'
  HEAD = 'HEAD'
  OPTIONS = 'OPTIONS'
  POST = 'POST'
  PUT = 'PUT'
}

const httpMethod = z.nativeEnum(HttpMethod)

interface Request {
  method: HttpMethod,
  url: string | URL,
  body?: object | string
}

type GetRequest = {
  method: HttpMethod.GET
  url: Request['url']
  body: never
}

const getRequest = z.object({
  method: z.literal(HttpMethod.GET),
  url: z.string(),
})

type PostRequest = {
  method: HttpMethod.POST
  url: Request['url']
  body: Request['body']
}

const postRequest = z.object({
  method: z.literal(HttpMethod.POST),
  url: z.string(),
  body: z.any(), // or something better
})

// etc
```

You might notice that I never reuse `httpMethod`. I build a zod parser for all the methods, but I can't then pull the individual parsers out of it. I can get the literals out with
```ts
method: z.literal(httpMethod.enum.GET)
```
but I can't get the _parsers_ out of it like this:
```ts
method: httpMethod.GET
```
so I end up having to redeclare all the parsers with literals (as you suggest).

There's no bug or problem. I was just hoping for a little extra _convenience_ for working with enums because I use them a lot.

**Reply by JacobWeisenburger** [reply]
unfortunately, zod doesn't create individual parsers for each value in the enum. It just loops over each value and checks if it matches the value passed to parse. I understand that it might be a little more convenient to do `httpMethod.GET` instead of `z.literal(httpMethod.enum.GET)`, but it just seems unnecessary. You're only saving a few characters. But maybe I'm misunderstanding what you are asking.

**Comment by JacobWeisenburger** [comment]
If you are looking for ways to save some characters, you can use the following:

```ts
enum HttpMethod {
    DELETE = 'DELETE',
    GET = 'GET',
    HEAD = 'HEAD',
    OPTIONS = 'OPTIONS',
    POST = 'POST',
    PUT = 'PUT',
}

const httpMethod = z.nativeEnum( HttpMethod )

const requestSchema = z.object( {
    method: httpMethod,
    url: z.string(),
    body: z.string().or( z.record( z.any() ) ).optional(),
} )
type Request = z.infer<typeof requestSchema>
// type Request = {
//     body?: string | Record<string, any> | undefined
//     method: HttpMethod
//     url: string
// }

const getRequest = requestSchema.extend( {
    method: z.literal( HttpMethod.GET ),
} ).omit( { body: true } )
type GetRequest = z.infer<typeof getRequest>
// type GetRequest = {
//     method: HttpMethod.GET
//     url: string
// }

const postRequest = requestSchema.extend( {
    method: z.literal( HttpMethod.POST ),
} )
type PostRequest = z.infer<typeof postRequest>
// type PostRequest = {
//     body?: string | Record<string, any> | undefined
//     method: HttpMethod.POST
//     url: string
// }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by jamesarosen** [reply]
```ts
const getRequest = requestSchema.extend( {
    method: z.literal( HttpMethod.GET ),
} ).omit( { body: true } )
type GetRequest = z.infer<typeof getRequest>
```
This is nice. It's not what I had been hoping for, but it certainly helps!

**Comment by JacobWeisenburger** [comment, marked as answer]
While working on something else I found `z.enum().extract` and remembered this discussion. Do you think it will help you?

```ts
const httpMethod = z.enum( [ 'DELETE', 'GET', 'HEAD', 'OPTIONS', 'POST', 'PUT' ] )

const requestSchema = z.object( {
    method: httpMethod,
    url: z.string(),
    body: z.string().or( z.record( z.any() ) ).optional(),
} )
type Request = z.infer<typeof requestSchema>
// type Request = {
//     method: "DELETE" | "GET" | "HEAD" | "OPTIONS" | "POST" | "PUT"
//     url: string
//     body?: string | Record<string, any> | undefined
// }

const getRequest = requestSchema.extend( {
    method: httpMethod.extract( [ 'GET' ] ),
} ).omit( { body: true } )
type GetRequest = z.infer<typeof getRequest>
// type GetRequest = {
//     method: "GET"
//     url: string
// }

const postRequest = requestSchema.extend( {
    method: httpMethod.extract( [ 'POST' ] ),
} )
type PostRequest = z.infer<typeof postRequest>
// type PostRequest = {
//     method: "POST"
//     url: string
//     body?: string | Record<string, any> | undefined
// }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by jamesarosen** [reply]
That's really nice! It's not the syntax I had asked for, but it has exactly the right "refinement" semantics. Great find 👏
